File: src/main/java/at/flauschigesalex/defaultLibrary/file/FileHandler.java
package at.flauschigesalex.defaultLibrary.file;

import lombok.Getter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.*;
import java.util.ArrayList;

@Getter
@SuppressWarnings({"UnusedReturnValue", "unused", "BooleanMethodIsAlwaysInverted"})
public final class FileHandler {

    private final File file;
    private JsonManager jsonManager;

    public FileHandler(final @NotNull String path) {
        this(new File(path));
    }
    public FileHandler(final @NotNull File file) {
        this.file = file;
    }

    public boolean createFile() {
        if (file.exists())
            return true;

        try {
            return file.createNewFile();
        } catch (final IOException fail) {
            fail.printStackTrace();
        }
        return false;
    }

    public boolean createJsonFile() {
        if (file.exists())
            return true;

        return createFile() && write("{}");
    }

    public boolean createDirectory() {
        if (file.exists())
            return true;

        return file.mkdir();
    }

    public boolean delete() {
        return purge(file);
    }

    public @Nullable InputStream readStream() {
        try {
            return new FileInputStream(file);
        } catch (final Exception ignore) {
        }
        return null;
    }

    public @Nullable String readString() {
        if (!isReadable())
            return null;

        final StringBuilder builder = new StringBuilder();
        final InputStream stream = readStream();
        if (stream != null) {
            try {
                for (byte nom : stream.readAllBytes())
                    builder.append((char) nom);
                return builder.toString();
            } catch (final IOException ignore) {
            }
        }
        return null;
    }

    public boolean write(final @NotNull Object object) {
        return this.write(object.toString().getBytes());
    }

    public boolean write(final @NotNull InputStream inputStream) {
        try {
            this.write(inputStream.readAllBytes());
        } catch (Exception ignore) {}
        return false;
    }

    public boolean write(final byte[] bytes) {
        if (!isWritable())
            return false;

        try {
            OutputStream stream = new FileOutputStream(file);
            stream.write(bytes);
            stream.close();
            return true;
        } catch (Exception fail) {
            fail.printStackTrace();
        }
        return false;
    }

    private boolean purge(final @NotNull File file) {
        if (!file.exists())
            return true;

        if (file.isDirectory()) {
            final File[] contents = file.listFiles();
            if (contents != null)
                for (File contentFile : contents)
                    if (!purge(contentFile))
                        return false;
        }
        return file.delete();
    }

    public boolean isReadable() {
        return file.exists() && file.isFile() && file.canRead();
    }

    public boolean isWritable() {
        return this.isReadable() && file.canWrite();
    }

    public String toString() {
        return getFile().getPath() + "/";
    }
}

File: src/main/java/at/flauschigesalex/defaultLibrary/file/FileHandler.kt
package at.flauschigesalex.defaultLibrary.file

import java.io.File
import java.io.FileOutputStream
import java.io.InputStream

@Suppress("unused", "MemberVisibilityCanBePrivate")
class FileHandler(val file: File) : DataHandler() {

    constructor(filePath: String) : this(File(filePath))

    fun createJsonFile(): File? {
        return this.createFile()?.apply {
            write("{}")
        }
    }
    fun createFile(): File? {
        if (file.exists())
            return file

        return tryCatch({
            if (file.createNewFile()) file else null
        })
    }
    fun createDirectory(): File? {
        if (file.exists() && file.isDirectory)
            return file

        return tryCatch({
            if (file.mkdir()) file else null
        })
    }

    override fun readStream(): InputStream? {
        return tryCatch({
            file.inputStream()
        })
    }

    fun write(obj: Any): Boolean {
        return write(obj.toString().toByteArray())
    }
    fun write(stream: InputStream): Boolean {
        return this.write(stream.readAllBytes())
    }
    fun write(bytes: ByteArray): Boolean {
        if (!this.isWritable())
            return false

        return tryCatch({
            FileOutputStream(file).apply {
                this.write(bytes)
                this.close()
            }
        }) != null
    }

    private fun delete(file: File, deep: Boolean = true): Boolean {
        if (!file.exists())
            return true

        if (file.isDirectory && deep) {
            val files = file.listFiles()
            if (files != null)
                for (dirFile in files)
                    if (!delete(dirFile))
                        return false
        }

        return file.delete()
    }

    fun delete(deep: Boolean = true): Boolean {
        return this.delete(file, deep)
    }

    override fun isReadable(): Boolean {
        return file.exists() && file.isFile && file.canRead()
    }

    override fun isWritable(): Boolean {
        return this.isReadable() && file.canWrite()
    }

    override fun toString(): String {
        return file.toString()
    }
}